File: ragSearch.js
const { ChromaClient } = require('chromadb');
const { pipeline } = require('@xenova/transformers');

// 全局单例
let embedder = null;
let collection = null;

// 初始化向量模型（只加载一次）
async function initEmbedder() {
  if (!embedder) {
    console.log('正在加载向量模型...');
    embedder = await pipeline('feature-extraction', 'Xenova/paraphrase-multilingual-MiniLM-L12-v2');
    console.log('向量模型加载完成。');
  }
  return embedder;
}

// 初始化 Chroma 集合
async function initCollection() {
  if (!collection) {
    const client = new ChromaClient({ path: './chroma_db' });
    collection = await client.getOrCreateCollection({ name: 'labor_law' });
    console.log('Chroma 集合已连接。');
  }
  return collection;
}

// 文本向量化（平均池化）
async function encodeTexts(texts) {
  const embedder = await initEmbedder();
  const embeddings = [];
  for (const text of texts) {
    const output = await embedder(text, { pooling: 'mean', normalize: true });
    embeddings.push(Array.from(output.data));
  }
  return embeddings;
}

// 检索函数
async function searchLaws(query, topK = 3) {
  const collection = await initCollection();
  const queryEmbedding = (await encodeTexts([query]))[0];

  const results = await collection.query({
    queryEmbeddings: [queryEmbedding],
    nResults: topK,
  });

  const laws = [];
  if (results.documents && results.documents[0]) {
    for (let i = 0; i < results.documents[0].length; i++) {
      laws.push({
        title: results.metadatas[0][i].title,
        content: results.documents[0][i],
        source: results.metadatas[0][i].source,
        tags: results.metadatas[0][i].tags,
      });
    }
  }
  return laws;
}

module.exports = { searchLaws };

File: rag_search.py
import sys
import json
from sentence_transformers import SentenceTransformer
import chromadb

# 初始化模型和向量库
model = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')
client = chromadb.PersistentClient(path="./chroma_db")
collection = client.get_collection("labor_law")

def search(query, top_k=3):
    query_embedding = model.encode(query).tolist()
    results = collection.query(
        query_embeddings=[query_embedding],
        n_results=top_k
    )
    laws = []
    if results['documents'] and results['documents'][0]:
        for i in range(len(results['documents'][0])):
            meta = results['metadatas'][0][i]
            laws.append({
                'title': meta['title'],
                'content': results['documents'][0][i],
                'source': meta['source'],
                'tags': meta.get('tags', '')
            })
    return laws

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "请提供查询语句"}, ensure_ascii=False))
        sys.exit(1)
    query = sys.argv[1]
    top_k = int(sys.argv[2]) if len(sys.argv) > 2 else 3
    result = search(query, top_k)
    print(json.dumps(result, ensure_ascii=False))